File: day23/typescript/day23.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

enum Status {
    running,
    terminated,
}

type Op = (number | string)[];
type Ops = { [name: string]: (state: State, op: Op) => State };
type Registers = { [name: string]: number };
interface State {
    pc: number;
    regs: Registers;
    status: Status;
}

const lookup = (state: State, x: number | string) => (
    isNaN(+x) ? R.pipe(R.path<number>(['regs', x]), R.defaultTo(0))(state) : +x
);

const ops: Ops = {
    set: (s, [x, y]) => ({ ...s, regs: { ...s.regs, [x]: lookup(s, y) }}),
    sub: (s, [x, y]) => ops.set(s, [x, lookup(s, x) - lookup(s, y)]),
    mul: (s, [x, y]) => ops.set(s, [x, lookup(s, x) * lookup(s, y)]),
    jnz: (s, [x, y]) => ({ ...s, pc: s.pc + (lookup(s, x) !== 0 ? lookup(s, y) - 1 : 0) }),
};

const perform = (state: State, code: string, size: number) => {
    const [op, ...ps] = R.split(' ', code);
    const s = ops[op](state, ps);
    const pc = s.pc + 1;
    return {
        ...s,
        pc,
        status: pc < 0 || pc >= size ? Status.terminated : s.status,
    };
};

const exec = (state: State) => (lines: string[]) => {
    let count = 0;
    while (state.status === Status.running) {
        if (true /*lines[state.pc].startsWith('mul')*/) {
            count++;
        }

        state = perform(state, lines[state.pc], lines.length);
    }

    return count;
};

const part1 = exec({ pc: 0, regs: {}, status: Status.running });

const isPrime = (num: number) => (
    R.all(n => num % n !== 0, R.range(2, Math.sqrt(num) + 1)) && num !== 1
);

const part2 = () => R.pipe<number, number[], number[], number>(
    R.times(n => 108100 + (n * 17)),
    R.filter<number>(n => !isPrime(n)),
    R.length,
)(1001);

(async () => {
    const input = await promisify(readFile)('day23/input.txt', 'utf8');
    const lines = R.split('\r\n', input);

    console.log(part1(lines));
    console.log(part2());
})();

File: day23/haskell/day23.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE RankNTypes #-}

import Data.Char (isAlpha)
import Data.Map (Map, empty)
import Data.Maybe (mapMaybe)
import Data.Monoid (Sum(..))
import qualified Data.Vector as V
import Control.Applicative (many)
import Control.Lens
import Control.Monad (when)
import Control.Monad.State (StateT, execStateT, get)
import Control.Monad.Writer (Writer, tell, execWriter)
import Control.Monad.Trans.Maybe (MaybeT, runMaybeT)

type Param = Either Char Int

data Op
    = Set Char Param
    | Sub Char Param
    | Mul Char Param
    | Jnz Param Param
      deriving (Eq, Show)

data Program = Program
    { _pc     :: Int
    , _regs   :: Map Char Int
    , _code   :: V.Vector Op
    } deriving (Show)
makeLenses ''Program

type App = MaybeT (StateT Program (Writer (Sum Int)))

parse :: String -> V.Vector Op
parse = V.fromList . mapMaybe (go . words) . lines
    where go ["set", x, y]  = Just $ Set (head x) (f y)
          go ["sub", x, y]  = Just $ Sub (head x) (f y)
          go ["mul", x, y]  = Just $ Mul (head x) (f y)
          go ["jnz", x, y]  = Just $ Jnz (f x) (f y)
          go _              = Nothing
          f [c] | isAlpha c = Left c
          f str             = Right $ read str

reg :: Char -> Lens' Program Int
reg r = regs . at r . non 0

val :: Param -> App Int
val = either (use . reg) pure

perform :: Op -> App ()
perform (Set x y) = (reg x .=) =<< val y
perform (Sub x y) = (reg x -=) =<< val y
perform (Mul x y) = (reg x *=) =<< val y
perform (Jnz x y) = (/= 0) <$> val x >>= flip when ((pc +=) =<< pred <$> val y)

step :: App ()
step = do
    c <- use pc
    Just o <- preuse (code . ix c)
    when (isMul o) (tell . Sum $ 1)
    perform o >> pc += 1
    where isMul (Mul _ _) = True
          isMul _         = False

execApp :: App a -> Program -> Writer (Sum Int) Program
execApp = execStateT . runMaybeT

part1 :: V.Vector Op -> Int
part1 = getSum . execWriter . execApp (many step) . Program 0 empty

part2 :: Int
part2 = length . filter (not . isPrime) $ [108100,108117..125100]
    where isqrt = floor . sqrt . fromIntegral
          isPrime k = null [ x | x <- [2..isqrt k], k `mod` x == 0 ]

main :: IO ()
main = do
    input <- parse <$> readFile "day23/input.txt"
    print . part1 $ input
    print part2